added return_grad for all types of rnnt loss

durson opened the pull request:
Grad tensors can be useful for the full (ordinary and pruned) rnnt losses.

durson committed:
added return_grad for all types of rnnt loss

pkufool commented:
Can you also fix this issue (https://github.com/k2-fsa/fast_rnnt/pull/25#issuecomment-1641313800), thanks!

durson committed:
lifted T >= S for regular case

durson commented:
I fixed [#25 (comment)](https://github.com/k2-fsa/fast_rnnt/pull/25#issuecomment-1641313800)
However I preserved S > 0 requirement since otherwise it is not compatible with torchaudio and unittest was not passing.

Summary of the changes:
* Added return_grad for all rnnt losses.
* Lifted T >= S requirement for regular case. That included making sure that shapes match not only for the whole tensor but for batch instances as well (testing boundaries if provided).
* Added s_range adjustment as per D.Povey suggestion and added unit test to check for correctness.
* Nitpicks - torch.Tensor -> Tensor, making bonudaries optional in appropriate cases.

pkufool commented on fast_rnnt/python/fast_rnnt/rnnt_loss.py:
```diff
@@ -657,28 +679,41 @@ def get_rnnt_prune_ranges(
+    if is_regular:
+        Ss = boundary[:, 2]
+        Ts = boundary[:, 3]
+        s_range_min = (Ss - 2).div(Ts, rounding_mode="trunc").add(3).max().item()
```
Could you explain this formula, I don't get you idea, sorry!

I think it should be `(s_range - 1) * Ts + 1 >= Ss + 1` and the simplified version is `s_range_min = Ss // Ts + 2`.  BTW, I checked several cases of both formula, the results are the same.

durson replied:
Thanks for checking that, your formula is simpler while still working correctly.

pkufool replied:
No, I think my formula won't work if `Ts== 1`,  you can double check that. Thanks!

pkufool replied:
Wait, It seems that your formula has a minimum `s_range_min` 3, which means we can't train a model with `s_range==2`. I think we can use `Ss // Ts + 2` and add a constraint `if s_range_min > S + 1: s_range_min = S + 1`.

durson replied:
Thanks, very nice catch. The correct formula is 
`(Ss - 1) // Ts + 2` and it works for all cases. I updated the code accordingly.

pkufool commented:
@durson  I have added the github actions to this repo, could you sync the PR with the latest master, thanks!

durson committed:
Merge remote-tracking branch 'upstream/master'

durson committed:
black reformat

durson committed:
black -l80 reformat

durson commented:
Merged latest master to fork and tests are passing.

durson committed:
fixed s_range adjustment rule

pkufool commented:
Thanks! merging! I would be very nice if you can also make a PR to https://github.com/k2-fsa/k2 (k2/python/k2/rnnt_loss.py).
